tests/realms.html

```diff
@@ -0,0 +1,342 @@
+                  'writable should be in method realm');
+  }, `${classname}\'s readable and writable attributes should come from the ` +
+     'same realm as the method definition');
+}
+
+function runTextEncoderTests() {
+  promise_test(async () => {
+    const objId = await constructAndStore('TextEncoder');
+    const writePromise = writeInWriteRealm(objId, 'A');
+    const result = await readInReadRealm(objId);
+    await writePromise;
+    assert_equals(result.constructor, constructorRealm.Object,
+                  'result should be in constructor realm');
+    assert_equals(result.value.constructor, constructorRealm.Uint8Array,
+                  'chunk should be in constructor realm');
+  }, 'the output chunks when read is called after write should come from the ' +
+     'same realm as the constructor of TextEncoder');
+
+  promise_test(async () => {
+    const objId = await constructAndStore('TextEncoder');
+    const chunkPromise = readInReadRealm(objId);
+    writeInWriteRealm(objId, 'A');
+    // Now the read() should resolve.
+    const result = await chunkPromise;
+    assert_equals(result.constructor, constructorRealm.Object,
+                  'result should be in constructor realm');
+    assert_equals(result.value.constructor, constructorRealm.Uint8Array,
+                  'chunk should be in constructor realm');
+  }, 'the output chunks when write is called with a pending read should come ' +
+     'from the same realm as the constructor of TextEncoder');
+
+  promise_test(async t => {
+    const objId = await constructAndStore('TextEncoder');
+    // Read first to relieve backpressure.
+    const readPromise = readInReadRealm(objId);
+    // promise_rejects() does not permit directly inspecting the rejection, so
+    // it's necessary to write it out long-hand.
+    try {
+      // Write an invalid chunk.
+      await writeInWriteRealm(objId, {
+        toString() { return {}; }
+      });
+      assert_unreached('write should fail');
+    } catch (err) {
+      assert_equals(err.constructor, constructorRealm.TypeError,
+                    'write TypeError should come from constructor realm');
+    }
+
+    try {
+      await readPromise;
+      assert_unreached('read should fail');
+    } catch (err) {
+      assert_equals(err.constructor, constructorRealm.TypeError,
+                    'read TypeError should come from constructor realm');
+    }
+  }, 'TypeError for unconvertable chunk should come from constructor realm ' +
+     'of TextEncoder');
+}
+
+function runTextDecoderTests() {
+  promise_test(async () => {
+    const objId = await constructAndStore('TextDecoder');
+    const writePromise = writeInWriteRealm(objId, new Uint8Array([65]));
+    const result = await readInReadRealm(objId);
+    await writePromise;
+    assert_equals(result.constructor, constructorRealm.Object,
+                  'result should be in constructor realm');
+    // Strings are always in the current realm, so no point in checking.
```

Comment:
Sorry. I'm not familiar with this point. Could you give me some pointer to the normative document to understand this?

Reply:
I've made the comment more specific of the exact behaviour I observe. @domenic, you know your way around the ECMAScript spec, do you know where this behaviour is defined?

Reply:
Strings as primitives do not have a realm at all.

However, they get implicitly converted from lowercase-S "string" primitives to an uppercase-S `String` object if you try to access properties of them, e.g. `"foo".constructor` or similar.

This implicit conversion always uses the current realm. It happens in https://tc39.github.io/ecma262/#sec-toobject but it doesn't explicitly state the realm, I think because in ES current is such an assumed default that nobody even thought to specify it.

Reply:
Got it. Thanks Domenic!